Respect tablet shard assignment

tgrabiec opened the pull request:
This PR changes the system to respect shard assignment to tablets in tablet metadata (system.tablets):
1. The tablet allocator is changed to distribute tablets evenly across shards taking into account currently allocated tablets in the system. Each tablet has equal weight. vnode load is ignored.
2. CDC subsystem was not adjusted (not supported yet)
3. sstable sharding metadata reflects tablet boundaries
5. resharding is NOT supported yet (the node will abort on boot if there is a need to reshard tablet-based tables)
6. The system is NOT prepared to handle tablet migration / topology changes in a safe way.
7. Sstable cleanup is not wired properly yet

After this PR, dht::shard_of() and schema::get_sharder() are deprecated. One should use table::shard_of() and effective_replication_map::get_sharder() instead.

To make the life easier, support was added to obtain table pointer from the schema pointer:

```
schema_ptr s;
s->table().shard_of(...)
```

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1317/

avikivity commented on schema/schema_registry.cc:
```diff
@@ -54,6 +54,19 @@ void schema_registry::init(const db::schema_ctxt& ctxt) {
+        // ignore
+    }
+}
+
```
I'm fairly negative on this. The schema should be just a schema. What are those contexts that only have a schema and don't have the database?

I'll try to suggest alternatives as I review.

tgrabiec replied:
It will also be useful to get rid of the "static props" thing added in this PR: https://github.com/scylladb/scylladb/pull/13170#issuecomment-1469848086

tgrabiec replied:
It's truly necessary only to add a check to schema::get_sharder() which catches sites which use it with tablet-based tables.

Other uses could either lookup in the database or propagate erm across layer.

There is some advantage in that we avoid table lookups, which would have to happen multiple times at different layers.

avikivity replied:
I believe most call sites can be adjusted to use an already-looked-up table. Or outright eliminated, like multishard_mutation_query. However I don't propose doing that as a prerequisite, we won't make progress.

> cdc/metadata.cc:    auto shard_id = dht::shard_of(shard_cnt, entry.sharding_ignore_msb, tok);


cdc sharding will be reworked for (and greatly benfit from) tablets

> compaction/compaction.cc:        auto shard = dht::shard_of(*_schema, dk.token());

This is for resharding compactions, which are unnecessary with compaction groups.

> cql3/statements/select_statement.cc:        unsigned shard = dht::shard_of(*_schema, key_ranges[0].start()->value().as_decorated_key().token());

has access to global context.

> db/commitlog/commitlog_replayer.cc:        auto shard = fm.shard_of(schema);

Can be delegated access to global context.


> db/size_estimates_virtual_reader.cc:            return shard_of(s, dht::get_token(s,

Is in table context

> db/virtual_table.cc:    return dht::shard_of(*_s, dk.token()) == this_shard_id();

Is in table context


> dht/i_partitioner.cc:sharder::shard_of(const token& t) const {
> dht/i_partitioner.cc:    return dht::shard_of(_shard_count, _sharding_ignore_msb_bits, t);
> dht/i_partitioner.cc:unsigned shard_of(const schema& s, const token& t) {
> dht/i_partitioner.cc:    return s.get_sharder().shard_of(t);
> dht/i_partitioner.cc:    auto shard = _range.start() ? _sharder.shard_of(_range.start()->value().token()) : token::shard_of_minimum_token();
> dht/i_partitioner.cc:    auto start_shard = sharder.shard_of(start_token);
> dht/i_partitioner.hh:unsigned shard_of(const schema&, const token&);
> dht/sharder.hh:            , _start_boundary(_sharder.shard_of(_start_token) == shard ?
> dht/token-sharding.hh:unsigned shard_of(unsigned shard_count, unsigned sharding_ignore_msb_bits, const token& t);
> dht/token-sharding.hh:    virtual unsigned shard_of(const token& t) const;
> dht/token.cc:shard_of(unsigned shard_count, unsigned sharding_ignore_msb_bits, const token& t) {
> dht/token.cc:    if (sharder.shard_of(t) != shard_idx) {


This is the infrastructure we're eliminating

> docs/dev/cdc.md:_entries[i][p.shard_of(tok, _entries[i].streams.size(), _entries[i].sharding_ignore_msb)]
> Binary file docs/kb/flamegraph.svg matches

Ignorable

> multishard_mutation_query.cc:            const auto shard = sharder.shard_of(rit->as_partition_start().key().token());
> multishard_mutation_query.cc:    const auto shard = sharder.shard_of(pkey.token());
> multishard_mutation_query.cc:    const auto shard = 
sharder.shard_of(compaction_state.partition_start.key().token());

Irrelevant for tablets

> mutation/frozen_mutation.hh:    unsigned shard_of(const schema& s) const {
> mutation/frozen_mutation.hh:        return dht::shard_of(s, dht::get_token(s, key()));
> mutation/mutation.hh:    unsigned shard_of() const {
> mutation/mutation.hh:        return dht::shard_of(*schema(), token());

To be eliminated

> mutation_writer/multishard_writer.cc:        return _s->get_sharder().shard_of(mf.as_partition_start().key().token());
> mutation_writer/shard_based_splitting_writer.cc:        _current_shard = dht::shard_of(*_schema, ps.key().token());
> readers/multishard.cc:    _current_shard = _sharder.shard_of(token);


Irrelevant for tablets


> replica/database.cc:        auto m_shard = dht::shard_of(*s, dht::get_token(*s, muts[i].key()));
> replica/database.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> replica/database.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> replica/database.hh://   use shard_of() for data
> replica/database.hh:    // All mutations must be owned by the current shard (in terms of dht::shard_of).
> replica/table.cc:        if (dht::shard_of(*s, pos.token()) != this_shard_id()) {

Have access to global context 

> scylla-gdb.py:def shard_of(ptr):
> scylla-gdb.py:        shard = shard_of(cache_region.impl())
> scylla-gdb.py:                if shard_of(int(desc.address)) != shard:

Irrelevant

> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:            unsigned shard = dht::shard_of(*schema, token);
> service/storage_proxy.cc:    return dht::shard_of(s, token);
> service/storage_proxy.cc:    if (coordinator_in_replica_set && dht::shard_of(*s, token) == this_shard_id()) {
> service/storage_proxy.cc:    auto shard = m.shard_of();
> service/storage_proxy.cc:    auto shard = m.shard_of(*s);
> service/storage_proxy.cc:    auto shard = m.shard_of(*s);
> service/storage_proxy.cc:    auto shard = fm.shard_of(*s);
> service/storage_proxy.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());
> service/storage_proxy.cc:        unsigned shard = dht::shard_of(*s, pr.start()->value().token());

Have access to global context

> test/boost/commitlog_test.cc:            return m.shard_of();
> test/boost/database_test.cc:    auto shard = m.shard_of();
> test/boost/database_test.cc:            auto shard = m.shard_of();
> test/boost/database_test.cc:                auto shard = m.shard_of();
> test/boost/database_test.cc:                auto shard = m.shard_of();
> test/boost/multishard_mutation_query_test.cc:                return smp::submit_to(dht::shard_of(*schema, mut.decorated_key().token()), [&env, gs = global_schema_ptr(schema), mut = freeze(mut)] () mutable {
> test/boost/multishard_mutation_query_test.cc:        const auto res = db.invoke_on(sharder.shard_of(pkey.token()), [gs = global_schema_ptr(s), &pkey, &slice] (replica::database& db) {
> test/boost/mutation_reader_test.cc:        auto shard = sharder.shard_of(start_token);
> test/boost/mutation_reader_test.cc:                    return remote_partitioner.shard_of(pkey.token()) == 0;
> test/boost/mutation_writer_test.cc:                    auto shard = s->get_sharder().shard_of(m.token());
> test/boost/mutation_writer_test.cc:                                        auto shard = sharder.shard_of(mf_opt->as_partition_start().key().token());
> test/boost/mutation_writer_test.cc:                                        auto shard = sharder.shard_of(mf_opt->as_partition_start().key().token());
> test/boost/partitioner_test.cc:        BOOST_REQUIRE_EQUAL(sharder.shard_of(lim), i % shards);
> test/boost/partitioner_test.cc:            BOOST_REQUIRE_EQUAL(sharder.shard_of(prev_token(lim)), (i - 1) % shards);
> test/boost/partitioner_test.cc:            BOOST_REQUIRE_EQUAL(sharder.shard_of(sharder.token_for_next_shard(lim, next_shard)), next_shard);
> test/boost/partitioner_test.cc:                auto start_shard = input_sharder.shard_of(range_shard->start()->value());
> test/boost/partitioner_test.cc:                auto end_shard = input_sharder.shard_of(range_shard->end()->value());
> test/boost/partitioner_test.cc:            auto mid_shard = input_sharder.shard_of(midpoint);
> test/boost/partitioner_test.cc:    auto last_shard = sharder.shard_of(last_token);
> test/boost/per_partition_rate_limit_test.cc:        unsigned local_shard = dht::shard_of(*sptr, dht::get_token(*sptr, pk.view()));
> test/boost/secondary_index_test.cc:                unsigned shard = m.shard_of();
> test/lib/cql_test_env.cc:        auto shard = dht::shard_of(*schema, dk._token);
> test/lib/dummy_sharder.cc:unsigned dummy_sharder::shard_of(const dht::token& t) const {
> test/lib/dummy_sharder.cc:        return shard_token > t && shard_of(shard_token) == shard;
> test/lib/dummy_sharder.hh:    virtual unsigned shard_of(const dht::token& t) const override;
> test/lib/key_utils.cc:                if (shard && *shard != dht::shard_of(*s, dkey.token())) {


Fixing up the tests will be painful

> tools/scylla-types.cc:            auto shard = dht::shard_of(vm["shards"].as<unsigned>(), vm["ignore-msb-bits"].as<unsigned>(), dk.token());

Will be broken by tablets (but replaces with a query into the tablets table).

avikivity commented on locator/tablets.hh:
```diff
@@ -148,6 +139,12 @@ public:
+    /// If there is no replica on a given host, returns nullopt.
+    /// If the topology is transitional, also considers the new replica set.
+    /// The old replica set is preferred in case of ambiguity.
+    std::optional<shard_id> get_shard(tablet_id, host_id) const;
```
Shouldn't the direction matter? reads shouldn't consider the new node during write_both_read_old.

tgrabiec replied:
Yes, transitional topology is not really handled properly yet anywhere.

avikivity commented on locator/tablet_sharder.hh:
```diff
@@ -0,0 +1,77 @@
+
+namespace locator {
+
+class tablet_sharder : public dht::sharder {
```
tablets are an alternative to vnodes, I think we should try to avoid also making them participate in old-style sharding. I'll read on to see if I can find a way to avoid this.

tgrabiec replied:
There are users which can stay uniform for vnodes and tablets, and I think there's value in that. If not dht::sharder, it would be something else.

avikivity commented on locator/tablet_sharder.hh:
```diff
@@ -0,0 +1,77 @@
+    tablet_sharder(const token_metadata& tm, table_id table)
+            : _tm(tm)
+            , _table(table)
+    { }
```
Shouldn't the lookup be done here?

tgrabiec replied:
tablet_sharder is instantiated with erm. In some cases we instantiate erm with token_metadata which is lacking info for a gicven table, temporarily. I would have to swallow the lookup error and use optionals to return the error in accessors.

avikivity commented on locator/tablets.cc:
```diff
@@ -202,6 +202,7 @@ size_t tablet_metadata::external_memory_usage() const {
 
 class tablet_effective_replication_map : public effective_replication_map {
     table_id _table;
+    tablet_sharder _sharder;
```
For this use case, tablet_sharder does not need to be a dht::sharder (and the lookup can be merged with the lookup for the host_id too).

tgrabiec replied:
Which use case do you have in mind? This is in order to implement erm::get_sharder(), which needs to be uniform for vnodes and tablets, so we need some common abstraction.

avikivity replied:
I look at effective_replication_map as a coordiantor-side thing. I know it's used on the replica too, but IMO it's a mistake.

As a thought exercise, we should be able to have replica-only nodes and coordiantor-only nodes, with the coordinator using data_dictionary and effective_replication_map and the replica nodes using replica:: things. We're very far from such isolation (and views muddle it, because a view makes a replica act as a mini-coordinator), but let's keep it in mind.

So, for coordination, we only need the shard when we look up the host to talk to it. But since we use e_r_m on the replica side too, we might need the sharder.

avikivity commented on sstables/sstables.cc:
```diff
@@ -1545,12 +1545,12 @@ populate_statistics_offsets(sstable_version_types v, statistics& s) {
-create_sharding_metadata(schema_ptr schema, const dht::decorated_key& first_key, const dht::decorated_key& last_key, shard_id shard) {
+create_sharding_metadata(schema_ptr schema, const dht::sharder& sharder, const dht::decorated_key& first_key, const dht::decorated_key& last_key, shard_id shard) {
     auto prange = dht::partition_range::make(dht::ring_position(first_key), dht::ring_position(last_key));
     auto sm = sharding_metadata();
```
This is overkill for tablets+compaction groups.

When we have both tablets and compaction groups, it's enough to set sharding metadata = [first key, last key]. This is because, even if the sstable ends up spanning 2^n tablets with n > 0, there will never be "holes" that we need to ignore like the current sharding metadata.

So, if we end up throwing away the tablet sharder, this can be replaced with a bool_class\<generate_fine_grained_sharding_metadata> or whatever (which the table can propagate down).

tgrabiec replied:
Yes, but this is necessary for incremental development. The node will not boot without this. Later, when sstables will span only a single tablet, this metadata will naturally reduce to [first key, last key].

avikivity commented on main.cc:
```diff
@@ -1383,6 +1380,9 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
+            supervisor::notify("loading tablet metadata");
+            ss.local().load_tablet_metadata().get();
+
             supervisor::notify("loading non-system sstables");
```
Queued

avikivity commented on replica/distributed_loader.cc:
```diff
@@ -451,11 +451,18 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, distr
+        erms.start(sharded_parameter([&global_table] {
+            return global_table->get_effective_replication_map();
+        })).get();
+        auto stop_erms = deferred_stop(erms);
```
This could be done in another way: instead of delegating to the sstable and asking into to look at the owner shards, we could intersect the [first, last] token range with the tablet map. This could tell us not only whether the sstable is owned by multiple shards (and therefore needs immediate splitting) but also whether it's owned by multiple tablets (and needs non-immediate splitting).

Options:
1. keep as is
2. abort() for now
3. move the checking code here

tgrabiec replied:
It won't work now when sstables are shared by many tables, since [first, last] range can overlap with tablets from other shards, even though the sstable itself contains only ranges of tablets from a single shard.

avikivity commented on replica/table.cc:
```diff
@@ -2787,7 +2788,9 @@ class compaction_group::table_state : public compaction::table_state {
     sstables::sstable_writer_config configure_writer(sstring origin) const override {
-        return _t.get_sstables_manager().configure_writer(std::move(origin));
+        auto cfg = _t.get_sstables_manager().configure_writer(std::move(origin));
+        cfg.erm = _t.get_effective_replication_map();
```
It's enough to write [first, last] (or indeed, drop the sharding metadata section for sstables owned by tablets, since with compaction groups they will have no holes).

avikivity replied:
I guess it generates a circular dependency, since compaction groups depend on fixing up sharding.

avikivity commented on multishard_mutation_query.cc:
```diff
@@ -753,7 +760,9 @@ future<typename ResultBuilder::result_type> do_query(
+    auto& table = db.local().find_column_family(s);
+    auto erm = table.get_effective_replication_map();
+    auto ctx = seastar::make_shared<read_context>(db, s, erm, cmd, ranges, trace_state, timeout);
 
```
Note: multishard_mutation_query would never be called if storage_proxy only ever sends range queries to one tablet.

tgrabiec replied:
Fixed in V2.

avikivity commented on schema/schema.cc:
```diff
@@ -152,6 +152,11 @@ const dht::i_partitioner& schema::get_partitioner() const {
+    auto t = maybe_table();
+    if (t && !t->uses_static_sharding()) {
+        // Use table()->get_effective_replication_map()->get_sharder() instead.
+        on_internal_error(dblog, format("Attempted to obtain static sharder for table {}.{}", ks_name(), cf_name()));
```
We'll want to make those unit tests work with tablets too, but okay for now.

avikivity commented on test/topology/test_tablets.py:
```diff
@@ -67,6 +67,9 @@ async def test_tablet_metadata_propagates_with_schema_changes_in_snapshot_mode(m
 
+    manager.driver_close()
+    await manager.driver_connect(server=servers[0])
+
```
Queued

avikivity commented on test/topology_experimental_raft/test_tablets.py:
```diff
@@ -100,8 +100,8 @@ async def test_scans(manager: ManagerClient):
-    rows = manager.cql.execute("SELECT * FROM test.test;")
-    assert len(list(rows)) == len(keys)
+    rows = manager.cql.execute("SELECT * FROM test.test;").all()
+    assert len(rows) == len(keys)
```
Queued

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1595,7 +1596,7 @@ class resharding_compaction final : public compaction {
     compaction_writer create_compaction_writer(const dht::decorated_key& dk) override {
-        auto shard = dht::shard_of(*_schema, dk.token());
+        auto shard = _table_s.get_table().shard_of(dk.token());
         auto sst = _sstable_creator(shard);
```
table_state abstracts away the concept of table. We need to add shard_of() method into table_state interface, rather than adding an accessor to underlying table. Unit tests for example might interact with compaction layer without actually having an actual table backing it.

tgrabiec replied:
There's no table_state implementation which doesn't have a table right now.

What's the reason for the abstraction? We end up reflecting most of a table provides.

raphaelsc replied:
we're abstracting the object being compacted with it. be it a table, or compaction group, or whatever. adding a table to table_state interface is a tremendous step back. we did lots of effort to remove direct table reference from the compaction layer. please add shard_of() to table_state instead, and whatever else you need.

raphaelsc replied:
By the way, table_state is a terrible name. It's the compaction layer view of compaction groups.

tgrabiec replied:
Every compaction object has a table, so it doesn't break the abstraction.I don't see how adding shard_of() is good, but adding table() is bad.

raphaelsc replied:
That's a layer violation. Now you have table_state (compaction group) -> table -> all compaction groups (including the one being identified by table_state), and compaction group can be "casted" into table_state. So you're also adding a cyclic dependence.

I am even planning to rename table_state to compaction_group. compaction_group is only a group of sstables which are eligible to be compacted together. It doesn't have to have a table backing it.

And I want compaction tests in the future to work without an actual table.

So please just add shard_of() to table_state.

raphaelsc replied:
Theoretically, not even shard_of() should belong to compaction_group. Compaction group should tell more things like if it owns a particular token. Which shard owns a particular token is indeed beyond its domain.

In perform_cleanup() interface in compaction manager, we have:

future<> perform_cleanup(owned_ranges_ptr sorted_owned_ranges, compaction::table_state& t);

So you can see that we provide sorted owned ranges in addition to the table state (group).

I can see that resharding is currently the sole user of shard_of(). So perhaps we can feed the function as a lambda param to resharding compaction.

raphaelsc replied:
We can potentially feed table::shard_of() into resharding with compaction_descriptor. See how reshard() in distributed_loader.cc feeds a sstable creator lambda into compaction.

raphaelsc replied:
> The kind of cycle you mention is already there: table -> compaction_group::table_state -> table.
> 
> table_state delegates to the table. Both implementations. This suggests that the layer separation at this level is artificial, in practice you want a table. In tests you also delegate to the table because otherwise you would have to reimplement the table. Abstracting this doesn't give us any benefit, just impediment for development.
> 

Memtables + sstables are now implemented in replica::compaction_group. So theoretically my test can implement table_state interface with c__g only, and not a replica::table. But I understand your point.

> > Theoretically, not even shard_of() should belong to compaction_group. So perhaps we can feed the function as a lambda param to resharding compaction.
> 
> Makes sense, I'll try that.

Thanks.

avikivity replied:
table_state::get_table() doesn't exist in the current code.

A lot of table_state's methods return the same thing for all compaction groups, while others will return different things. This suggests that the actual state is in the second class of methods, while the first class describes configuration, not state.

```c++
// config
    virtual const schema_ptr& schema() const noexcept = 0;
    // min threshold as defined by table.
    virtual unsigned min_compaction_threshold() const noexcept = 0;
    virtual bool compaction_enforce_min_threshold() const noexcept = 0;
    virtual sstables::compaction_strategy& get_compaction_strategy() const noexcept = 0;
    virtual reader_permit make_compaction_reader_permit() const = 0;
    virtual sstables::sstables_manager& get_sstables_manager() noexcept = 0;
    virtual sstables::shared_sstable make_sstable() const = 0;
    virtual sstables::sstable_writer_config configure_writer(sstring origin) const = 0;
    virtual bool is_auto_compaction_disabled_by_user() const noexcept = 0;
    virtual bool tombstone_gc_enabled() const noexcept = 0;
// state
    virtual const sstables::sstable_set& main_sstable_set() const = 0;
    virtual const sstables::sstable_set& maintenance_sstable_set() const = 0;
    virtual std::unordered_set<sstables::shared_sstable> fully_expired_sstables(const std::vector<sstables::shared_sstable>& sstables, gc_clock::time_point compaction_time) const = 0;
    virtual const std::vector<sstables::shared_sstable>& compacted_undeleted_sstables() const noexcept = 0;
    virtual compaction_strategy_state& get_compaction_strategy_state() noexcept = 0;
    virtual api::timestamp_type min_memtable_timestamp() const = 0;
    virtual future<> on_compaction_completion(sstables::compaction_completion_desc desc, sstables::offstrategy offstrategy) = 0;
    virtual const tombstone_gc_state& get_tombstone_gc_state() const noexcept = 0;
    virtual compaction_backlog_tracker& get_backlog_tracker() = 0;
    virtual const std::string& get_group_id() const noexcept = 0;
    virtual seastar::condition_variable& get_staging_done_condition() noexcept = 0; // ?
```

avikivity commented on locator/load_sketch.hh:
```diff
@@ -0,0 +1,91 @@
+    using shard_id = seastar::shard_id;
+    struct shard_load {
+        shard_id id;
+        size_t load;
```
What are the units? tablets or bytes? If bytes, should be uint64_t since size_t is local RAM sized.

tgrabiec replied:
>What are the units?

tablets.

raphaelsc commented:
> > To make the life easier, support was added to obtain table pointer from the schema pointer:
> > ```
> > schema_ptr s;
> > s->table().shard_of(...)
> > ```
> 
> That introduces a dependency loop, no? schema is a basic vocabulary type and is meaningful everywhere (replica/coordiantor/data dictionary). Tables are a replica-side thing only.

Agree. This introduces a bad cyclic dependence between the two. Table is aware of its schema, but not the opposite. I know schema stores the table's id, but that's as far as it goes. Referencing the table object itself from schema may not be a good decising looking forward.

I looked at one of the places, which uses the table accessor in schema, and it seems we could have alternatively looked up the table from the database, referenced by the query_processor. You can potentially introduce a helper to database class, to compute the shard of a token from a schema. So it reduces the boilerplate for you.

avikivity commented on locator/load_sketch.hh:
```diff
@@ -0,0 +1,91 @@
+    }
+};
+
+} // namespace locator
```
test?

avikivity replied:
^

tgrabiec replied:
Missed this, sorry.

tgrabiec replied:
> Please enhance the explanations in the patch changelogs.

Any particular patch?

avikivity commented on locator/load_sketch.hh:
```diff
@@ -0,0 +1,91 @@
+        std::pop_heap(n._shards.begin(), n._shards.end(), shard_load_cmp());
+        shard_load& s = n._shards.back();
+        auto shard = s.id;
+        s.load += 1;
```
So the units are tablets; please add a comment.

tgrabiec replied:
Ok.

avikivity commented on locator/load_sketch.hh:
```diff
@@ -0,0 +1,91 @@
+public:
+    load_sketch(const token_metadata& tm, std::optional<host_id> host = std::nullopt)
+            : _topo(tm.get_topology()) {
+        for (auto&& [table, tmap] : tm.tablets().all_tables()) {
```
This looks like a stall in the making.

tgrabiec replied:
Will fix.

avikivity commented:
Back to the question of the circular dependency, where low-level things like sstables begin to depend on high-level things like effective_replication_map. I think it's possible to avoid these dependencies, but only after we integrate compaction groups, and compaction group integration depends on this series. It's also a non-mechanical transformation, so it's better to approach each site individually rather than in a big series.

So I'm okay with merging this, but I'd like to start work on removing the schema->table link and sstable->effective_replication_map dependency soon.

raphaelsc commented:
> Back to the question of the circular dependency, where low-level things like sstables begin to depend on high-level things like effective_replication_map. I think it's possible to avoid these dependencies, but only after we integrate compaction groups, and compaction group integration depends on this series. It's also a non-mechanical transformation, so it's better to approach each site individually rather than in a big series.
> 
> So I'm okay with merging this, but I'd like to start work on removing the schema->table link and sstable->effective_replication_map dependency soon.

Yes. Storage aware of distribution is asking for trouble later on.

tgrabiec commented:
In V2:
  - Added more docs to tablet_sharder
  - Make tablet_sharder::next_shard() be consistent with shard_of() and return shard 0 for tokens missing locally
  - Fix infinite loop in tablet_sharder::token_for_next_shard()
  - Dropped table_state::table()
  - Propagate sharder to resharding compaction via compaction descriptor
  - Documented load_sketch (class comment and shard_load::load unit)
  - Avoid reactor stall in load_sketch population by extracting to populate()
  - Avoid range merging on coordinator side for tablets during scan to avoid multishard reader:
    500762946d storage_proxy: Avoid multishard reader for tablets
  - Fixed split_range_to_single_shard() to work with tablet sharding:
    d3d398b758 dht: Make split_range_to_single_shard() prepared for tablet sharder
  - Attach table pointer to static schemas:
    c47d948fba schema_registry: Make learn(schema_ptr) attach entry to the target schema

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1548/

avikivity commented on schema/schema.hh:
```diff
@@ -826,6 +827,14 @@ public:
+
+    // Like maybe_table() but throws replica::no_such_column_family if the table is not set.
+    replica::table& table() const;
+
```
It's really like to avoid this. A schema isn't a table.

Unfortunately shard_of deeply embedded in contexts where we don't have any global knowledge, like mutation::shard_of(). So it will be a long project to adjust all shard queries to refer to global state instead of schema dependent state. But I don't see a way to avoid it now.

denesb replied:
Don't we already have a concept of a sharder? Can't we use it to abstract how shard assignment works?

tgrabiec replied:
> Don't we already have a concept of a sharder? Can't we use it to abstract how shard assignment works?

Yes, that's what this series does. What do you propose to change?

tgrabiec replied:
You can ease the pain by reframing it like this. schema_ptr represents a table schema version. Each table schema version refers to a particular physical table, by its id(). table() is a materialization of that relationship. 

It's similar to how a prepared select statement stores a schema_ptr instead of table name.

avikivity replied:
It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.

The current schema::get_sharder() says: "this is the algorithm and parameters used to determine s key's shard".

After the change, it no longer refers to an algorithm and parameters, because there is lots of mutable state indirectly referred to. The schema is no longer a value, it contains references.

So, we should work to get rid of get_sharder(). It's lots of work so it will have to happen later.

tgrabiec replied:
> It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.

I get your point about having schema which doesn't have to be associated with a table, but that's not what our current schema object is. It is a table schema, which is associated with a table. It doesn't own the table, it refers to it. There is `database::find_column_family(schema_ptr)` which allows one to lookup the table from schema_ptr. If we remove table_id from it, we need to pass around not schema_ptr but a new object:

```
struct table_schema {
    table_id id;
    raw_schema_ptr s;
}
```

This refactor is orthogonal to the argument, because then we would be arguing about having table() inside this new table_schema object.

tgrabiec replied:
Regardless of the table() thing, I agree we should phase out uses of `schema::get_sharder()`.

avikivity replied:
> > It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.
> 
> I get your point about having schema which doesn't have to be associated with a table, but that's not what our current schema object is. It is a table schema, which is associated with a table. It doesn't own the table, it refers to it. There is `database::find_column_family(schema_ptr)` which allows one to lookup the table from schema_ptr. If we remove table_id from it, we need to pass around not schema_ptr but a new object:
> 
> ```
> struct table_schema {
>     table_id id;
>     raw_schema_ptr s;
> }
> ```
> 
> This refactor is orthogonal to the argument, because then we would be arguing about having table() inside this new table_schema object.

Maybe I can rephrase the point about table_id as "schema{,_ptr} is misnamed". But then it becomes an exercise in arguing semantics, let's not do that.

So the conclusion is to drop schema::get_sharder(). Probably it will be easier once we wean tablets off all multishard things. Or maybe not, and we just need to decorate effective_replication_map with the sharder and change all those APIs to get an effective_replication_map_ptr (or a replica-side equivalent). That will also serve to prevent from the mapping changing under our feet.

tgrabiec commented:
In V3:
  - fix CI failure due to missing include in load_sketch.hh header which trips `ninja checkheaders`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1571/

avikivity commented:
Please enhance the explanations in the patch changelogs.

avikivity commented:
> > Please enhance the explanations in the patch changelogs.
> 
> Any particular patch?

As an example, "dht: Take sharder externally in splitting functions". Why is it a good idea? If I later see a patch "dht: Don't take sharder externally in splitting functions", why shouldn't I apply it?

Please check that all patches document why, not just what.

tgrabiec commented:
In V4:
  - Added unit test for load_sketch
  - Improved patch descriptions

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1581/

kbr-scylla commented:
Dequeued, breaks `cdc_test.py::TestCdc::test_add_field_with_cdc[Multi_DC_cluster]` and `cdc_test.TestCdc.test_change_field_type_with_cdc[Multi_DC_cluster]`
e.g. https://jenkins.scylladb.com/view/nexts/job/scylla-master/job/next/6079/testReport/junit/cdc_test/TestCdc/Tests___Sanity_Tests___test_add_field_with_cdc_Multi_DC_cluster_/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1648/

tgrabiec commented:
In V5:

  - Fixed failure in cdc_test due to a bug in schema_registry::get_or_load() which would miss to attach the table if it is already concurrently being loaded, and as a result no_such_column_family being thrown
  - Added more unit tests
  - Added get_feature_service() to cql_test_env
  - Extracted utils::throttle out of row_cache_test

avikivity commented:
Please rebase, github indicates a conflict which might not be trivial.

tgrabiec commented:
In V5.1:
   - rebased

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1671/

tgrabiec commented:
In V5.2:

 - fixed debug-mode compilation of row_cache_test due to missed namespace qualifier after class move

tgrabiec commented:
CI failed due to https://github.com/scylladb/scylladb/issues/14136

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1679/

avikivity commented:
Meanwhile, lots more conflicts appeared.

tgrabiec commented:
In V5.2:
  - rebased

avikivity commented:
Looks like the unit tests hang.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1683/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1695/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1769/

tgrabiec commented:
CI failure due to https://github.com/scylladb/scylladb/issues/14217

tgrabiec commented:
In V5.3:
  - rebased, maser has a fix for https://github.com/scylladb/scylladb/issues/14217

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1882/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1898/

avikivity commented:
Please rebase. I tried to merge but I get failures in test_tablets.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1920/

tgrabiec commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1920/

https://github.com/scylladb/scylladb/issues/14303

raphaelsc commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1920/
> 
> #14303

If the only issue caught in CI is known, and surely not related to this series, I guess we can ignore CI output and just merge the branch into next. It's not clear to me what's the consensus among the maintainers.

tgrabiec commented:
In this case, the test added by the PR catches the issue, so we would introduce flakiness (high rate) to the build.

raphaelsc commented:
> In this case, the test added by the PR catches the issue, so we would introduce flakiness (high rate) to the build.

Ah that issue was just opened ~25 min ago. Means that you wrote a really good test :-)

tgrabiec commented:
In V6:
  - rebased (conflicts with fencing)
  - added a fix for fencing bug https://github.com/scylladb/scylladb/issues/14303 which trips tablets_test.py
  - added shard count to the list of logged node attributes in existing logger invocations

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1930/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1962/
